src/client/datascience/jupyter/kernels/kernelSelections.ts

```diff
@@ -200,24 +199,61 @@ export class InstalledJupyterKernelSelectionListProvider
-                // If we have a default kernel name and a non-absolute path just hide the item
-                // Otherwise we end up showing a bunch of "Python 3 - python" default items for
-                // other interpreters
-                const match = detectDefaultKernelName(item.name);
-                if (match) {
-                    return path.isAbsolute(item.path);
-                }
-                return true;
-            })
-            .map((item) => getQuickPickItemForKernelSpec(item, this.pathUtils));
+        const selections = await Promise.all(
+            items
+                .filter((item) => {
+                    // If we have a default kernel name and a non-absolute path just hide the item
+                    // Otherwise we end up showing a bunch of "Python 3 - python" default items for
+                    // other interpreters
+                    const match = detectDefaultKernelName(item.name);
+                    if (match) {
+                        // Check if this is a kernel we registerd in the old days.
+                        // If it is, then no need to display that (selecting kernels registered is done by selecting the corresponding interpreter).
+                        // Hence we can hide such kernels.
+                        // Kernels we create will end with a uuid (with - stripped), & will have interpreter info in the metadata.
+                        if (
+                            item.metadata?.interpreter &&
+                            item.name.length > 32 &&
+                            item.name.slice(-32).toLowerCase() === item.name
+                        ) {
+                            return false;
+                        }
+
+                        // If we have the interpreter information this kernel belongs to and the kernel has custom env
+                        // variables, then include it in the list.
+                        if (item.interpreterPath && item.env) {
+                            return true;
+                        }
+                        // Else include it only if the path is available for the kernel.
+                        return path.isAbsolute(item.path);
+                    }
+                    return true;
+                })
+                .map((item) => getQuickPickItemForKernelSpec(item, this.pathUtils))
+                .map(async (item) => {
+                    // Ensure we have the associated interpreter information.
+                    const selection = item.selection as ReadWrite<KernelSpecConnectionMetadata>;
+                    if (selection.interpreter || !isPythonKernelConnection(selection)) {
+                        return item;
+                    }
+                    selection.interpreter = await this.kernelService.findMatchingInterpreter(
+                        selection.kernelSpec,
+                        cancelToken
+                    );
+                    item.selection = selection;
```

Comment:
Is this line necessary?

Reply:
Yes, to update it with a new item.

Reply:
Maybe I need to read up on javascript then. Isn't this a reference to the item.selection object?

```typescript
const selection = item.selection
```

So if you change selection.foo, it also changes item.selection.foo?

Reply:
No it does behave as I expected. Object assignment is by reference unless you explicitly clone/copy in the assignment.

Reply:
Oh shoot, you are right, i don't need to reset the variable, I've done that in `getKernelSelections` (few lines above, line 181).
Agreed, this is not necessary.